pyquil/api/_base_connection.py

```diff
@@ -180,3 +183,350 @@ def validate_run_items(run_items):
+    """REST payload for :py:func:`ForestConnection._expectation`"""
+    if operator_programs is None:
+        operator_programs = [Program()]
+
+    if not isinstance(prep_prog, Program):
+        raise TypeError("prep_prog variable must be a Quil program object")
+
+    payload = {'type': TYPE_EXPECTATION,
+               'state-preparation': prep_prog.out(),
+               'operators': [x.out() for x in operator_programs]}
+
+    if random_seed is not None:
+        payload['rng-seed'] = random_seed
+
+    return payload
+
+
+def qvm_run_payload(quil_program, classical_addresses, trials, needs_compilation, isa,
+                    measurement_noise, gate_noise, random_seed):
+    """REST payload for :py:func:`ForestConnection._qvm_run`"""
+    if not quil_program:
+        raise ValueError("You have attempted to run an empty program."
+                         " Please provide gates or measure instructions to your program.")
+    if not isinstance(quil_program, Program):
+        raise TypeError("quil_program must be a Quil program object")
+    validate_run_items(classical_addresses)
+    if not isinstance(trials, integer_types):
+        raise TypeError("trials must be an integer")
+    if needs_compilation and not isa:
+        raise TypeError("ISA cannot be None if program needs compilation preprocessing.")
+
+    payload = {"type": TYPE_MULTISHOT,
+               "addresses": list(classical_addresses),
+               "trials": trials}
+    if needs_compilation:
+        payload["uncompiled-quil"] = quil_program.out()
+        payload["target-device"] = {"isa": isa.to_dict()}
+    else:
+        payload["compiled-quil"] = quil_program.out()
+
+    if measurement_noise is not None:
+        payload["measurement-noise"] = measurement_noise
+    if gate_noise is not None:
+        payload["gate-noise"] = gate_noise
+    if random_seed is not None:
+        payload['rng-seed'] = random_seed
+
+    return payload
+
+
+def qpu_run_payload(quil_program, classical_addresses, trials, needs_compilation, isa):
+    """REST payload for :py:func:`ForestConnection._qpu_run_async`"""
+    if not quil_program:
+        raise ValueError("You have attempted to run an empty program."
+                         " Please provide gates or measure instructions to your program.")
+
+    if not isinstance(quil_program, Program):
+        raise TypeError("quil_program must be a Quil program object")
+    validate_run_items(classical_addresses)
+    if not isinstance(trials, integer_types):
+        raise TypeError("trials must be an integer")
+
+    payload = {"type": TYPE_MULTISHOT,
+               "addresses": list(classical_addresses),
+               "trials": trials}
+
+    if needs_compilation:
+        payload["uncompiled-quil"] = quil_program.out()
+        if isa:
+            payload["target-device"] = {"isa": isa.to_dict()}
+    else:
+        payload["compiled-quil"] = quil_program.out()
+
+    def _wrap_program(self, program):
```

Comment:
I don't see how this could ever get called?

Reply:
good spot